### PR TITLE
Enable Widget Tests and fixes for rendering of nested Output Widgets

### PR DESCRIPTION
**Pull request by DonJayamanne**
Also added/enabeld some much needed IPyWidget tests

### CODE REVIEW

File: src/kernels/execution/cellExecutionMessageHandler.ts [hunk 1]
```diff
@@ -72,33 +79,13 @@ function getParentHeaderMsgId(msg: KernelMessage.IMessage): string | undefined {
     return undefined;
 }
 
-/**
- * The Output Widget in Jupyter can render multiple outputs. However some of them
- * like vendored mimetypes cannot be handled by it.
- */
-function canMimeTypeBeRenderedByWidgetManager(mime: string) {
```
Comment (DonJayamanne) [inline src/kernels/execution/cellExecutionMessageHandler.ts:79]: Moved, as Widget manager cannot properly render Output Widgets either.

File: src/kernels/execution/cellExecutionMessageHandler.ts [hunk 2]
```diff
@@ -594,9 +651,17 @@ export class CellExecutionMessageHandler implements IDisposable {
 
         // We have some new outputs, that need to be placed immediately after the widget and before any other output
         // that doesn't belong to the widget.
-        const clearWidgetOutput = this.outputsAreSpecificToAWidget?.clearOutputOnNextUpdateToOutput === true;
-        if (this.outputsAreSpecificToAWidget) {
-            this.outputsAreSpecificToAWidget.clearOutputOnNextUpdateToOutput = false;
+        const clearOption = 'clearOutput' in options ? options.clearOutput : false;
```
Comment (DonJayamanne) [inline src/kernels/execution/cellExecutionMessageHandler.ts:654]: Bulk of he change is to turn a property into an array

File: src/kernels/execution/cellExecutionMessageHandler.ts [hunk 3]
```diff
@@ -58,6 +60,11 @@ type ExecuteResult = nbformat.IExecuteResult & {
 type DisplayData = nbformat.IDisplayData & {
     transient?: { display_id?: string };
 };
+type WidgetData = {
```
Comment (DonJayamanne) [inline src/kernels/execution/cellExecutionMessageHandler.ts:63]: Was defined inline in the code, moved it out as its used in two places

File: src/kernels/execution/cellExecutionMessageHandler.ts [hunk 4]
```diff
@@ -627,6 +692,9 @@ export class CellExecutionMessageHandler implements IDisposable {
         if (outputsAfterWidget.length === 0 && outputToAppend) {
             // No need to replace everything, just append what we need.
             task?.appendOutput(outputToAppend, cell).then(noop, noop);
+        } else if (clearWidgetOutput && !outputToAppend && outputsAfterWidget.length === 0) {
+            // We're supposed to clear the output after a specific widget, but there are no other outputs after the widget.
```
Comment (DonJayamanne) [inline src/kernels/execution/cellExecutionMessageHandler.ts:696]: Minor optimization, previously we were re-rendering the same output even when there were no changes.

File: src/test/datascience/widgets/standard.vscode.common.test.ts [hunk 5]
```diff
@@ -33,488 +33,642 @@ import { WidgetRenderingTimeoutForTests } from './constants';
...
-        traceInfo('Suite Setup VS Code Notebook - Execution');
+[false].forEach((useCDN) => {
+    /**
+     * Testing the following permutations:
+     * - VSCode Web + Remote Jupyter + CDN
+     * - VSCode Web + Remote Jupyter + No CDN
+     * - Desktop + Remote Jupyter + CDN
+     * - Desktop + Remote Jupyter + No CDN
+     * - Desktop + Local Kernel + CDN
+     * - Desktop + Local Kernel + No CDN
```
Comment (DonJayamanne) [inline src/test/datascience/widgets/standard.vscode.common.test.ts:44]: Ensure we run the tests in more permutations (previously the tests would only run with CDN, now they'll run with & without CDN in all environments - web, desktop & remote)

File: src/test/datascience/widgets/thirdpartyWidgets.vscode.common.test.ts [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+import {
+    assertOutputContainsHtml,
+    clickWidget,
+    executeCellAndDontWaitForOutput,
+    executeCellAndWaitForOutput,
+    initializeNotebookForWidgetTest
+} from './standardWidgets.vscode.common.test';
+
+/* eslint-disable @typescript-eslint/no-explicit-any, no-invalid-this */
+[true, false].forEach((useCDN) => {
```
Comment (DonJayamanne) [inline src/test/datascience/widgets/thirdpartyWidgets.vscode.common.test.ts:31]: Created a separate test file for 3rd party widgets

File: src/test/datascience/widgets/standardWidgets.vscode.common.test.ts [hunk 7]
```diff
@@ -32,19 +32,74 @@ import { initializeWidgetComms, Utils } from './commUtils';
...
+    await comms.click(cell, selector);
+}
+
 /* eslint-disable @typescript-eslint/no-explicit-any, no-invalid-this */
-suite('IPyWisdget Tests', function () {
-    const templateRootPath: Uri =
-        workspace.workspaceFolders && workspace.workspaceFolders.length > 0
-            ? urlPath.joinPath(workspace.workspaceFolders[0].uri, 'widgets', 'notebooks')
-            : Uri.file('');
+suite('Standard IPyWidget Tests', function () {
```
Comment (DonJayamanne) [inline src/test/datascience/widgets/standardWidgets.vscode.common.test.ts:96]: Moved 3rd party widgets into a seprate test file

File: src/test/datascience/widgets/thirdpartyWidgets.vscode.common.test.ts (hunk 6 above)
Comment (DonJayamanne) [reply src/test/datascience/widgets/thirdpartyWidgets.vscode.common.test.ts]: & enabled some of the tests that were previously disabled.

File: src/kernels/execution/cellExecutionMessageHandler.ts [hunk 8]
```diff
@@ -138,7 +125,19 @@ export class CellExecutionMessageHandler implements IDisposable {
...
+     * The way it works is:
+     * - Jupyter sends a comm open message indicating there's an output widget
+     * - Jupyter sends a comm message indicating all messages from now on belong to the output widget.
+     *      - Now if we have another output widget thats nested in the previous, then we get another comm open message.
+     *      - Jupyter sends a comm message indicating all messages from now on belong to the second output widget.
+     *      - Jupyter sends a comm message indicating all messages from now on don't belong to the second the output widget is done (comm message with msg_id = '').
+     * - Jupyter sends a comm message indicating all messages from now on don't belong to the first the output widget is done (comm message with msg_id = '').
+     *
+     * Basically its a stack, as we create new outputs nested within another we get comm messages and we need to keep track of the stack.
+     */
```
Comment (sadasant) [inline src/kernels/execution/cellExecutionMessageHandler.ts:139]: Very cool! Thank you for writing this 🙏

File: src/kernels/execution/cellExecutionMessageHandler.ts [hunk 9]
```diff
@@ -398,6 +398,20 @@ export class CellExecutionMessageHandler implements IDisposable {
             this.execution.executionOrder = msg.content.execution_count;
         }
     }
+    private handleCommOpen(msg: KernelMessage.ICommOpenMsg) {
+        this.ownedCommIds.add(msg.content.comm_id);
+        // Check if this is a message for an Output Widget,
+        const state: { _model_module: string } | undefined =
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            typeof msg.content.data === 'object' && msg.content.data ? (msg.content.data['state'] as any) : undefined;
```
Comment (sadasant) [inline src/kernels/execution/cellExecutionMessageHandler.ts:406]:
```suggestion
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const state: { _model_module: string } | undefined = msg.content.data?.state || undefined;
```

I’m wondering if this makes sense.

File: src/kernels/execution/cellExecutionMessageHandler.ts [hunk 10]
```diff
@@ -539,6 +562,37 @@ export class CellExecutionMessageHandler implements IDisposable {
...
+     * However some of them like the Output Widget & vendored mimetypes cannot be handled by it.
+     */
+    private canMimeTypeBeRenderedByWidgetManager(outputItem: NotebookCellOutputItem) {
+        const mime = outputItem.mime;
+        if (
+            mime == CellOutputMimeTypes.stderr ||
+            mime == CellOutputMimeTypes.stdout ||
+            mime == CellOutputMimeTypes.error
+        ) {
+            // These are plain text mimetypes that can be rendered by the Jupyter Lab widget manager.
```
Comment (sadasant) [inline src/kernels/execution/cellExecutionMessageHandler.ts:577]: Reads really well! Thank you

File: src/kernels/execution/cellExecutionMessageHandler.ts [hunk 11]
```diff
@@ -418,13 +432,17 @@ export class CellExecutionMessageHandler implements IDisposable {
...
+                    this.outputsAreSpecificToAWidget.push({
                         handlingCommId: msg.content.comm_id,
                         msgIdsToSwallow: data.state.msg_id
-                    };
-                } else if (this.outputsAreSpecificToAWidget?.handlingCommId === msg.content.comm_id) {
+                    });
+                } else if (
+                    this.outputsAreSpecificToAWidget.length &&
+                    this.outputsAreSpecificToAWidget[this.outputsAreSpecificToAWidget.length - 1].handlingCommId ===
+                        msg.content.comm_id
```
Comment (sadasant) [inline src/kernels/execution/cellExecutionMessageHandler.ts:442]:
```suggestion
                    this.outputsAreSpecificToAWidget.at(-1)?.handlingCommId ===
                        msg.content.comm_id
```

I was wondering if `at` was a thing yet or not, and seems like it is: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/at looks pretty cool

Comment (sadasant) [reply src/kernels/execution/cellExecutionMessageHandler.ts:442]: Ohh it depends on TypeScript using es2022

File: src/test/datascience/widgets/thirdpartyWidgets.vscode.common.test.ts [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+        await executeCellAndWaitForOutput(cell3, comms);
+        await assertOutputContainsHtml(cell1, comms, ['Continuous Slider']);
+        await assertOutputContainsHtml(cell2, comms, ['Continuous Text', '<input type="number']);
+        await assertOutputContainsHtml(cell3, comms, ['Continuous Slider', '>0<', '>123.00']);
+
+        // Update the textbox widget (for slider).
+        await comms.setValue(cell2, '.widget-text input', '5255');
+        await assertOutputContainsHtml(cell3, comms, ['>5255<', '>5378.0']);
+    });
+    test.skip('Render ipyvolume (slider, color picker, figure)', async function () {
```
Comment (rchiodo) [inline src/test/datascience/widgets/thirdpartyWidgets.vscode.common.test.ts:146]: How come this one is skipped?

**Review by rchiodo — APPROVED** [review]

**Review by IanMatthewHuff — APPROVED** [review]

File: src/test/datascience/widgets/thirdpartyWidgets.vscode.common.test.ts (hunk 12 above)
Comment (DonJayamanne) [reply src/test/datascience/widgets/thirdpartyWidgets.vscode.common.test.ts:146]: Was already skipped, will need to figure out why it wasn't working.

File: src/kernels/execution/cellExecutionMessageHandler.ts (hunk 9 above)
Comment (DonJayamanne) [reply src/kernels/execution/cellExecutionMessageHandler.ts:406]: Damn it, missed this, will add this in the next PR

**Pull request merged**